### PR TITLE
List bootspec entries in bootctl and use the default for kexec

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by keszybz** [comment]
Updated to drop the mempool commit.

File: src/shared/bootspec.c [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+        free(entry->device_tree);
+}
+
+int boot_entry_load(const char *path, BootEntry *entry) {
+        _cleanup_fclose_ FILE *f = NULL;
+        unsigned line = 1;
+        _cleanup_(boot_entry_free) BootEntry tmp = {};
+        int r;
+
+        f = fopen(path, "r");
```
Comment (poettering) [inline src/shared/bootspec.c]: "re"

File: src/shared/bootspec.c [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+                        return log_error_errno(r, "%s:%u: Line too long", path, line);
+                if (r < 0)
+                        return log_error_errno(r, "%s:%u: Error while reading: %m", path, line);
+
+                line++;
+
+                if (IN_SET(*strstrip(buf), '#', '\0'))
+                        continue;
+
+                p = strchr(buf, ' ');
```
Comment (poettering) [inline src/shared/bootspec.c:83]: maybe use extract_first_word?

File: src/shared/bootspec.c [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
```
Comment (poettering) [inline src/shared/bootspec.c:163]: as above: extract_first_word()?

File: src/shared/bootspec.c [hunk 4]
```diff
@@ -274,6 +275,71 @@ int boot_entries_find(const char *dir, BootEntry **entries, size_t *n_entries) {
...
+                        free_and_replace(entries[i].show_title, s);
+                }
+
+        if (!find_nonunique(entries, n_entries, arr))
+                return 0;
+
+        /* Add file name to non-unique titles */
+        for (i = 0; i < n_entries; i++)
+                if (arr[i]) {
+                        r = asprintf(&s, "%s (%s)", boot_entry_title(entries + i), entries[i].filename);
```
Comment (poettering) [inline src/shared/bootspec.c:338]: not that it matters, but strjoin() is a bazillion times faster than asprintf() if all you want to do is concat a few strings

File: src/shared/bootspec.c [hunk 5]
```diff
@@ -403,3 +408,201 @@ int boot_entries_load_config(const char *esp_path, BootConfig *config) {
...
+                *ret_pstart = pstart;
+        if (ret_psize)
+                *ret_psize = psize;
+        if (ret_uuid)
+                *ret_uuid = uuid;
+
+        return 0;
+}
+
+int find_esp(char **path,
```
Comment (poettering) [inline src/shared/bootspec.c:580]: i am not a fan of library functions that log on their own beyond LOG_DEBUG, unless this is very clear from their context or name. i.e. call this find_esp_and_warn() or so?

File: src/shared/bootspec.c (hunk 1 above)
Comment (keszybz) [reply src/shared/bootspec.c]: Ack.

File: src/shared/bootspec.c (hunk 4 above)
Comment (keszybz) [reply src/shared/bootspec.c:338]: True, but in this case speed is not important, and I find printing with a format string always much more readable than concatenation.

File: src/shared/bootspec.c (hunk 2 above)
Comment (keszybz) [reply src/shared/bootspec.c:83]: I don't think so. The problem is that this must implement the exact same semantics as the efi code. So any support for quoting or escapes is counterproductive. When changed to use extract_first_word, the code is the same length, but there's the potential that all the bell and whistles of extract_first_word will cause some misparsing. I prefer to keep this as is.

**Comment by keszybz** [comment]
Updated.

File: src/systemctl/systemctl.c [hunk 6]
```diff
@@ -143,6 +144,7 @@ static const char *arg_kill_who = NULL;
 static int arg_signal = SIGTERM;
 static char *arg_root = NULL;
 static usec_t arg_when = 0;
+static char *arg_esp_path;
```
Comment (poettering) [inline src/systemctl/systemctl.c]: this really should be initialized to NULL. While on Linux is it safe not to do this, valgrind does notice, and ANSI C isn't fine with this either. Also, I think it's more readable if we initialize global variables explicitly, and we do that for all others too

File: src/systemctl/systemctl.c [hunk 7]
```diff
@@ -3497,6 +3499,71 @@ static int prepare_firmware_setup(void) {
...
+        else if (pid == 0) {
+
+                const char* const args[] = {
+                        KEXEC,
+                        "--load", kernel,
+                        "--append", options,
+                        initrd ? "--initrd" : NULL, initrd,
+                        NULL };
+
+                /* Child */
```
Comment (poettering) [inline src/systemctl/systemctl.c:3560]: as a matter for cargo-culting i usually try to add the following to somewhere between fork() and execve() in a child:

```c
                (void) reset_all_signal_handlers();
                (void) reset_signal_mask();
                assert_se(prctl(PR_SET_PDEATHSIG, SIGTERM) == 0);
```
At least the first two we should probably do everywhere, no?

**Comment by keszybz** [comment]
`find_esp was already` renamed, or rather when it was moved, it was moved without the logging part, and a new `find_esp_and_warn` was added that calls `find_esp` and does the logging.

Updated with the suggested changes, and a few more commits on top to allow tests in `test/` to run automatically on Fedora when the kernel is unerneath `/boot/efi/`.

**Comment by keszybz** [comment]
Ah, I wrote that I implemented suggested changes, when in fact I did not:
> mkosi at least doesn't actually use the .conf snippets, but generates "unified" kernel images, that consist of the kernel itself, the initrd a root cmdline and so on all in one. sd-boot can also read those, and i figure we should support that here too, so that bootctl output is actually complete?

I'll work on that in a separate PR.

**Pull request merged**